src/sensors/elec.rs

```diff
@@ -14,21 +26,142 @@ pub fn filter_pde(pde: f32) -> bool {
+    x1*x2 + y1*y2 + z1*z2
+}
+
+// Dot product separation for now, need to implement other algorithm(s)
+fn dot_product_clusters(hits: &[MCQT], xyzs: &SensorMap, threshold: f32, min_sensors: usize) -> Option<(Vec<MCQT>, Vec<MCQT>)> {
+    let max_sens = hits.iter().max_by_key(|e| e.q).map(|e| e.sensor_id)?;
+    let &(xm, ym, zm) = xyzs.get(&max_sens)?;
+    let (c1, c2): (Vec<MCQT>, Vec<MCQT>) = hits.iter()
+            .filter(|&MCQT { q, .. }| *q as f32 > threshold)
+            .partition(|&MCQT { sensor_id, .. }| dot((xm, ym, zm), *xyzs.get(sensor_id).unwrap()) > Area::ZERO);
+    if c1.len() >= min_sensors && c2.len() >= min_sensors { Some((c1, c2)) } else { None }
+}
+
+/// Reconstruct LORs based on the barycentre of the 2 selected clusters (dot prod) without DOI.
+fn lor_from_sensor_positions(hits: &[MCQT], xyzs: &SensorMap, threshold: f32, min_sensors: usize) -> Option<Hdf5Lor> {
+    let (c1, c2) = dot_product_clusters(hits, xyzs, threshold, min_sensors)?;
+    // TODO time just taken as minimum of minima. Averaging used elsewhere.
+    let b1 = sipm_charge_barycentre(&c1, xyzs)?;
+    let b2 = sipm_charge_barycentre(&c2, xyzs)?;
+    let q1 = c1.iter().map(|&MCQT { q, .. }| q as f32).sum();
+    let q2 = c2.iter().map(|&MCQT { q, .. }| q as f32).sum();
+    // Just dummy time for now until we work out the best way to sort.
+    Some(Hdf5Lor {
+        dt: ns_(b2.3 - b1.3),
+        x1: mm_(b1.0), y1: mm_(b1.1), z1: mm_(b1.2),
+        x2: mm_(b2.0), y2: mm_(b2.1), z2: mm_(b2.2),
+        q1, q2, E1: f32::NAN, E2: f32::NAN,
+    })
+}
+
+fn sipm_charge_barycentre(hits: &[MCQT], xyzs: &SensorMap) -> Option<(Length, Length, Length, Time)> {
+    let mut qs = Ratio::ZERO;
+    let mut xx = Length::ZERO;
+    let mut yy = Length::ZERO;
+    let mut zz = Length::ZERO;
+    let mut tt = Time::ZERO;
+    for &MCQT{ sensor_id, q, t, .. } in hits {
+        let &(x, y, z) = xyzs.get(&sensor_id)?;
+        let q = ratio(q as f32);
+        qs += q;
+        xx += x * q;
+        yy += y * q;
+        zz += z * q;
+        if t < tt { tt = t};
+    }
+    Some((xx / qs, yy / qs, zz / qs, tt))
+}
+
+
+pub fn lor_reconstruction(first_file: &PathBuf, pde: f32, time_mu: f32, time_sigma: f32, threshold: f32, min_sensors: usize) -> Box<dyn Fn(&PathBuf) -> hdf5::Result<(Vec<Hdf5Lor>, usize)>> {
+    let time_smear = gaussian_sampler(time_mu, time_sigma);
+    // Need to do better than unwrap!
+    let xyzs = read_sensor_map(&first_file).unwrap();
+    Box::new(move |filename: &PathBuf| -> hdf5::Result<(Vec<Hdf5Lor>, usize)> {
+        let sensor_hits = read_sensor_hits(&filename, Bounds::none())?;
+        let detected_hits: Vec<MCSensorHit> = sensor_hits.iter()
+                                                       .filter(|_| filter_pde(pde))
+                                                       .map(|&MCSensorHit { event_id, sensor_id, time }| MCSensorHit { event_id, sensor_id, time: time + time_smear() })
```

Comment:
+ Break the line after `=` and dedent the rest accordingly.
+ Inside the `map` use [struct update syntax](https://doc.rust-lang.org/book/ch05-01-defining-structs.html#creating-instances-from-other-instances-with-struct-update-syntax)

So you'll end up with something like
```rust
        let detected_hits: Vec<MCSensorHit> =
            sensor_hits.iter()
                       .filter(|_| filter_pde(pde))
                       .map(|hit| MCSensorHit { time: hit.time + time_smear(), ..*hit })
                       .collect();
```